src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,375 @@
+        old_model_forward = self.lora_model.model.forward
+
+        def new_model_forward(*args, **kwargs) -> None:
+            # =========================== Forward pass with "dummy" scalings ==================
+
+            dummy_scalings = self.internal_xlora_classifier.make_dummy_scalings(*args, **kwargs)
+
+            for layer in all_layers:
+                layer.scalings = dummy_scalings
+
+            with torch.no_grad():
+                self.lora_model.disable_adapters()
+
+                try:
+                    scaling_pass_kwargs = kwargs.copy()
+                    scaling_pass_kwargs["output_hidden_states"] = True
+                    scaling_pass_kwargs["return_dict"] = True
+                    try:
+                        base_output = old_model_forward(*args, **scaling_pass_kwargs)
+                    finally:
+                        # Clean everything up
+                        for layer in all_layers:
+                            layer.scalings = None
+                finally:
+                    self.lora_model.enable_adapters()
+
+            xlora_scalings = self.internal_xlora_classifier(result=base_output, *args, **kwargs)
+
+            # =========================== Real forward pass with calculated scalings ==================
+
+            for layer in all_layers:
+                layer.scalings = xlora_scalings
+
+            try:
+                output = old_model_forward(*args, **kwargs)
+            finally:
+                # Clean everything up
+                for layer in all_layers:
+                    layer.scalings = None
+            return output
+
+        self.lora_model.model.forward = new_model_forward
+
+        n_classes = len(peft_config.adapters)
+        xlora_classifier = XLoraClassifier(model, peft_config, n_classes, total_swapped, device)
+
+        # Setup the model internal state
+        self.internal_xlora_classifier = xlora_classifier
+        self.internal_xlora_scalings = None  # type: ignore
+
+    def _maybe_freeze_all_adapters(self):
+        self.eval()
+        if not self.xlora_config.use_trainable_adapters:
+            for name, param in self.named_parameters():
+                if "lora_" in name:
+                    param.requires_grad = False
+
+    def generate(self, *args, **kwargs):
+        res = self.lora_model.generate(*args, **kwargs)  # type: ignore
+        #  This is necessary because we use PeftModel.disable_adapter() which reenables the adapters
+        self._maybe_freeze_all_adapters()
+        return res
+
+    def __getattr__(self, name: str):
+        """Forward missing attributes to the wrapped module."""
+        try:
+            return super().__getattr__(name)  # defer to nn.Module's logic
+        except AttributeError:
+            return getattr(self.lora_model, name)
+
+    def _mark_only_adapters_as_trainable(self, model: nn.Module) -> None:
```

Comment:
This whole method is never called during the tests. Can we delete it or are the tests missing an important aspect?

Reply:
I think we can remove it, we also don't want to be activating adapter if the config doesn't say so.